### PR TITLE
Let the ? operator work natively in try_stream!.

### PR DESCRIPTION
**Pull request by goffrie**
Insteads of desugaring `?` in the macro, we can have the async block
itself return `Result<(), E>`, and adjust the supporting code so that
`?` just works. The benefit is that this allows `?` operators that are
hidden behind macros.

### CODE REVIEW

**Review by taiki-e — APPROVED** [review]
LGTM, thanks!

**Comment by albel727** [comment]
Is this supposed to break code that compiled before? Because it does for me. Before this PR one could use "return;" in the stream body to stop yielding. Now this fails with 
```
expected type `()`
              found enum `std::result::Result<(), _>`
```

**Comment by albel727** [comment]
Ah yes, now that I read the PR with some coffee, of course it's supposed to break code. Well, simply returning Ok(()) is not a big deal, and the ability to return an Err is more flexible too. But I would really want a v0.3.1 release of async_stream, which lifts the Unpin on Stream::Item requirement, but with this PR included it would probably have to wait for a v0.4.

**Comment by goffrie** [comment]
Ah, that's a good point. I didn't realize it, but yes this is a breaking change.

I'll leave it to @taiki-e (or other maintainer :slightly_smiling_face:) to decide whether to cut a 0.4 or revert this PR.

**Comment by taiki-e** [comment]
@albel727 good catch!

I tend to accept this at 0.4, but for now, I'd like to revert this. (and I'd like to add more tests before re-land in 0.4.)

**Comment by SergioBenitez** [comment]
What is the difference between `try_stream!` and `stream`, besides the expected type of the `expr` in `yield expr`? If you take a `try_stream!` and replace every `yield expr` with `yield Ok(expr)`, can you simply swap the `try_stream!` invocation for a `stream!` invocation?

If the answer is yes, then this leads me to ask: are both `stream` and `try_stream` necessary, and should `?` _just work_ in `stream!`? If the answer is no, the docs should be updated to reflect this. Consider that these two functions appears to work identically:

```rust
fn f<S>(stream: S) -> impl Stream<Item = io::Result<u8>>
    where S: Stream<Item = io::Result<&'static str>>
{
    try_stream! {
        for await s in stream {
            let num = s?.parse();
            let num = num.map_err(|e| io::Error::new(io::ErrorKind::Other, e))?;
            yield num;
        }
    }
}

fn g<S>(stream: S) -> impl Stream<Item = io::Result<u8>>
    where S: Stream<Item = io::Result<&'static str>>
{
    stream! {
        for await s in stream {
            let num = s?.parse();
            let num = num.map_err(|e| io::Error::new(io::ErrorKind::Other, e))?;
            yield Ok(num);
        }
    }
}
```

**Comment by Kestrer** [comment]
? cannot be used inside `stream!` because the async block must evaluate to (). We could modify `stream!` to support evaluating to `Result<()>` in which case the difference with `try_stream!` would be that it requires an `Ok(())` at the end (ok-wrapping the `yield`s can be done by `AsyncStream`). However, that wouldn't work well with the Try trait as we'd have overlapping impls of when the async block evaluates to () or `T: Try`.

If `Try` was implemented for `()` I think the two macros could be unified however. I haven't seen any plans for that though.

**Comment by SergioBenitez** [comment]
> ? cannot be used inside `stream!` because the async block must evaluate to ().

I don't understand. My example code above uses `?` in `stream!` and works exactly as the `try_stream!` version does. Can you post an example where they would work differently?

**Comment by Kestrer** [comment]
Ah, right - it works because `?` in `stream!` expands to a `match` where the `Err` branch yields and returns. So yes, I think you're right - other than native `?` expansion and `Ok`-wrapping `stream!` and `try_stream!` are identical.

**Pull request merged**
